src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBarService/VsInfoBarService.cs

```diff
@@ -0,0 +1,132 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.ProjectSystem.VS.UI.InfoBarService;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.PackageRestore.InfoBarService;
+
+/// <summary>
+/// Implementation of <see cref="IInfoBarService"/> that pushes messages to the info bar attached to Visual Studio's main window.
+/// </summary>
+[Export(typeof(IInfoBarService))]
+internal partial class VsInfoBarService : IInfoBarService
+{
+    private readonly IProjectThreadingService _threadingService;
+    private readonly IVsShellServices _vsShellServices;
+    private readonly IVsUIService<SVsShell, IVsShell> _vsShell;
+    private readonly IVsUIService<SVsInfoBarUIFactory, IVsInfoBarUIFactory> _vsInfoBarFactory;
+    private readonly List<InfoBarEntry> entries = new();
+
+    [ImportingConstructor]
+    public VsInfoBarService(
+        IProjectThreadingService threadingService,
+        IVsShellServices vsShellServices,
+        IVsUIService<SVsShell, IVsShell> vsShell,
+        IVsUIService<SVsInfoBarUIFactory, IVsInfoBarUIFactory> vsInfoBarFactory)
+    {
+        _threadingService = threadingService;
+        _vsShellServices = vsShellServices;
+        _vsShell = vsShell;
+        _vsInfoBarFactory = vsInfoBarFactory;
+    }
+
+    public async Task ShowInfoBarAsync(string message, ImageMoniker image, CancellationToken cancellationToken, params InfoBarUI[] items)
+    {
+        Requires.NotNullOrEmpty(message, nameof(message));
+        Requires.NotNull(items, nameof(items));
+
+        if (!_vsShellServices.IsInServerMode)
```

Comment:
>if (!_vsShellServices.IsInServerMode) [](http://example.com/codeflow?start=8&length=37)

What is the right behavior if this is in server mode? Right now, it silently succeeds.

Reply:
Are you sure IsInServerMode works off the ui thread? This is invoked on a arbitrary thread in cycle detector.

Reply:
IsInServerMode -> is this to account for Nexus based scenarios? Do we need to explicitly call out a NotSupportedException when IsInServerMode = true?

Reply:
This was originally to account for Nexus scenarios, yes, and as we don't really have a server SKU under active development it probably doesn't matter much. I think it's reasonable to throw an exception if we are in server mode, and also agree with @jacdavis that we need to verify `IsInServerMode` is ok to call from an arbitrary thread, and/or transition to the UI thread.